File: daily-coding-challenge/2026/july/js/03_DatabaseMigration#327.js
/*

Database Migration

Given two database objects, return the second object with any missing properties from the first filled in.
  Fields that already exist in the record should not be overwritten.

Tests:
  migrateRecord({ username: "", posts: 0 }, { verified: true }) should return { username: "", posts: 0, verified: true }.
  migrateRecord({ username: "", posts: 0 }, { username: "camper", posts: 5 }) should return { username: "camper", posts: 5 }.
  migrateRecord({ username: "", posts: 0, verified: false }, { username: "camper" }) should return { username: "camper", posts: 0, verified: false }.
  migrateRecord({ username: "", posts: 0 }, { username: "camper", role: "admin" }) should return { username: "camper", role: "admin", posts: 0 }.
  migrateRecord({ username: "", email: "", posts: 0, verified: false, role: "user", banned: false }, { username: "camper", email: "camper@freecodecamp.org", role: "admin" }) should return { username: "camper", email: "camper@freecodecamp.org", role: "admin", posts: 0, verified: false, banned: false }.

*/

function migrateRecord(schema, record) {
  return { ...schema, ...record };
}

console.log(migrateRecord({ username: "", posts: 0 }, { verified: true })); // { username: "", posts: 0, verified: true }.
console.log(
  migrateRecord({ username: "", posts: 0 }, { username: "camper", posts: 5 }),
); // { username: "camper", posts: 5 }.
console.log(
  migrateRecord(
    { username: "", posts: 0, verified: false },
    { username: "camper" },
  ),
); // { username: "camper", posts: 0, verified: false }.
console.log(
  migrateRecord(
    { username: "", posts: 0 },
    { username: "camper", role: "admin" },
  ),
); // { username: "camper", role: "admin", posts: 0 }.
console.log(
  migrateRecord(
    {
      username: "",
      email: "",
      posts: 0,
      verified: false,
      role: "user",
      banned: false,
    },
    { username: "camper", email: "camper@freecodecamp.org", role: "admin" },
  ),
); // { username: "camper", email: "camper@freecodecamp.org", role: "admin", posts: 0, verified: false, banned: false }.

File: daily-coding-challenge/2026/july/python/03_DatabaseMigration#327.py
"""

Database Migration

Given two database objects, return the second object with any missing properties from the first filled in.
  Fields that already exist in the record should not be overwritten.

Tests:
  migrate_record({ "username": "", "posts": 0 }, { "verified": True }) should return { "username": "", "posts": 0, "verified": True }.
  migrate_record({ "username": "", "posts": 0 }, { "username": "camper", "posts": 5 }) should return { "username": "camper", "posts": 5 }.
  migrate_record({ "username": "", "posts": 0, "verified": False }, { "username": "camper" }) should return { "username": "camper", "posts": 0, "verified": False }.
  migrate_record({ "username": "", "posts": 0 }, { "username": "camper", "role": "admin" }) should return { "username": "camper", "role": "admin", "posts": 0 }.
  migrate_record({ "username": "", "email": "", "posts": 0, "verified": False, "role": "user", "banned": False }, { "username": "camper", "email": "camper@freecodecamp.org", "role": "admin" }) should return { "username": "camper", "email": "camper@freecodecamp.org", "role": "admin", "posts": 0, "verified": False, "banned": False }.

"""


def migrate_record(schema, record):
    return {**schema, **record}


print(
    migrate_record({"username": "", "posts": 0}, {"verified": True})
)  # { "username": "", "posts": 0, "verified": True }.
print(
    migrate_record({"username": "", "posts": 0}, {"username": "camper", "posts": 5})
)  # { "username": "camper", "posts": 5 }.
print(
    migrate_record(
        {"username": "", "posts": 0, "verified": False}, {"username": "camper"}
    )
)  # { "username": "camper", "posts": 0, "verified": False }.
print(
    migrate_record(
        {"username": "", "posts": 0}, {"username": "camper", "role": "admin"}
    )
)  # { "username": "camper", "role": "admin", "posts": 0 }.
print(
    migrate_record(
        {
            "username": "",
            "email": "",
            "posts": 0,
            "verified": False,
            "role": "user",
            "banned": False,
        },
        {"username": "camper", "email": "camper@freecodecamp.org", "role": "admin"},
    )
)  # { "username": "camper", "email": "camper@freecodecamp.org", "role": "admin", "posts": 0, "verified": False, "banned": False }.
